shared/Microsoft.Extensions.ObjectMethodExecutor.Sources/ObjectMethodExecutorAwaitable.cs

```diff
@@ -84,6 +90,25 @@ public void OnCompleted(Action continuation)
             {
                 _onCompletedMethod(_customAwaiter, continuation);
             }
+
+            public void UnsafeOnCompleted(Action continuation)
+            {
+                // If the underlying awaitable implements ICriticalNotifyCompletion, use its UnsafeOnCompleted.
+                // If not, fall back on using its OnCompleted.
+                //
+                // Why this is safe:
+                // - Implementing ICriticalNotifyCompletion is a way of saying the caller can choose whether it
+                //   needs the execution context to be preserved (which it signals by calling OnCompleted), or
+                //   that it doesn't (which it signals by calling UnsafeOnCompleted). Obviously it's faster *not*
+                //   to preserve and restore the context, so we prefer that where possible.
+                // - If a caller doesn't need the execution context to be preserved and hence calls UnsafeOnCompleted,
+                //   there's no harm in preserving it anyway - it's just a bit of wasted cost. That's what will happen
+                //   if a caller sees that the proxy implements ICriticalNotifyCompletion but the proxy chooses to
+                //   pass the call on to the underlying awaitable's OnCompleted method.
```

Comment:
@anurse @davidfowl Are you happy that this big comment is accurate? We should attempt to have some explanation of what we're doing here, because it's far from obvious that it's valid otherwise.